src/mongo/db/auth/external/external_sasl_authentication_session.cpp

```diff
@@ -0,0 +1,272 @@
+    Foundation.
+
+    Percona Server for MongoDB is distributed in the hope that it will
+    be useful, but WITHOUT ANY WARRANTY; without even the implied
+    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+    See the GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public
+    License along with Percona Server for MongoDB.  If not, see
+    <http://www.gnu.org/licenses/>.
+======= */
+
+#define MONGO_LOG_DEFAULT_COMPONENT ::mongo::logger::LogComponent::kAccessControl
+
+#include <sasl/sasl.h>
+
+#include "mongo/db/auth/native_sasl_authentication_session.h"
+
+#include <boost/scoped_ptr.hpp>
+#include <boost/range/size.hpp>
+
+#include "mongo/base/init.h"
+#include "mongo/base/status.h"
+#include "mongo/base/string_data.h"
+#include "mongo/bson/util/bson_extract.h"
+#include "mongo/client/sasl_client_authenticate.h"
+#include "mongo/db/commands.h"
+#include "mongo/db/auth/authorization_manager.h"
+#include "mongo/db/auth/authorization_manager_global.h"
+#include "mongo/db/auth/authorization_session.h"
+#include "mongo/db/auth/sasl_options.h"
+#include "mongo/db/auth/external/external_sasl_authentication_session.h"
+#include "mongo/util/assert_util.h"
+#include "mongo/util/log.h"
+#include "mongo/util/mongoutils/str.h"
+#include "mongo/util/net/sock.h"
+
+namespace mongo {
+
+    using boost::scoped_ptr;
+
+namespace {
+
+    SaslAuthenticationSession* createExternalSaslAuthenticationSession(
+        AuthorizationSession* authzSession,
+        const std::string& mechanism) {
+        return new IntegratedAuthenticationSession(authzSession);
+    }
+
+    // External SASL session factory needs to be set AFTER the native one has been set.
+    // TODO: Use this in case other non-external tests fail:: MONGO_INITIALIZER_GENERAL(ExternalSaslServerCore, ("PreSaslCommands"), ("PostSaslCommands"))
+    MONGO_INITIALIZER_WITH_PREREQUISITES(ExternalSaslServerCore, ("NativeSaslServerCore"))
+        (InitializerContext* context) {
+        if (saslGlobalParams.hostName.empty())
+            saslGlobalParams.hostName = getHostNameCached();
+        if (saslGlobalParams.serviceName.empty())
+            saslGlobalParams.serviceName = "mongodb";
+
+        int result = sasl_server_init(NULL, saslDefaultServiceName);
+        if (result != SASL_OK) {
+            log() << "Failed Initializing External Auth Session" << std::endl;
+            return ExternalSaslAuthenticationSession::getInitializationError(result);
+        }
+
+        log() << "Initialized External Auth Session" << std::endl;
+        SaslAuthenticationSession::create = createExternalSaslAuthenticationSession;
+        return Status::OK();
+    }
+
+} //namespace
+
+    ExternalSaslAuthenticationSession::ExternalSaslAuthenticationSession(
+        AuthorizationSession* authzSession) :
+        NativeSaslAuthenticationSession(authzSession),
+        _saslConnection(NULL),
+        _mechanism("") {
+    }
```

Comment:
Will initialize char * and length appropriately and poison SASL result code with explicit SASL error code.